Setting up audit sinks can block unseal

Vault version:
```
root@vault-1:~# vault -version
Vault v0.6.2
```

config:
```
{
  "backend": {
    "consul": {
      "address": "127.0.0.1:8500", 
      "path": "vault"
    }
  }, 
  "disable_mlock": true, 
  "listener": {
    "tcp": {
      "address": "0.0.0.0:8200", 
      "tls_disable": 1
    }
  }
}
```

My cluster looks like:
```
$ lxc list
+-----------------+---------+-----------------------+-------------------------------------------+------------+-----------+
|      NAME       |  STATE  |         IPV4          |                   IPV6                    |    TYPE    | SNAPSHOTS |
+-----------------+---------+-----------------------+-------------------------------------------+------------+-----------+
| consul-server-1 | RUNNING | 10.170.131.180 (eth0) | <redacted> (eth0) | PERSISTENT | 0         |
+-----------------+---------+-----------------------+-------------------------------------------+------------+-----------+
| consul-server-2 | RUNNING | 10.170.131.200 (eth0) | <redacted> (eth0) | PERSISTENT | 0         |
+-----------------+---------+-----------------------+-------------------------------------------+------------+-----------+
| consul-server-3 | RUNNING | 10.170.131.244 (eth0) | <redacted> (eth0) | PERSISTENT | 0         |
+-----------------+---------+-----------------------+-------------------------------------------+------------+-----------+
| vault-1         | RUNNING | 10.170.131.69 (eth0)  | <redacted> (eth0) | PERSISTENT | 0         |
+-----------------+---------+-----------------------+-------------------------------------------+------------+-----------+
| vault-2         | RUNNING | 10.170.131.94 (eth0)  | <redacted> (eth0) | PERSISTENT | 0         |
```

Consul config:
```
root@consul-server-1:~# cat /etc/consul.d/config.json 
{"bind_addr": "0.0.0.0", "bootstrap_expect": 3, "data_dir": "/var/consul", "datacenter": "main", "enable_debug": true, "encrypt": "", "log_level": "info", "retry_interval": "15s", "retry_join": [], "server": true, "ui": true}
root@consul-server-1:~# consul members
Node             Address              Status  Type    Build  Protocol  DC
consul-server-1  10.170.131.180:8301  alive   server  0.7.0  2         main
consul-server-2  10.170.131.200:8301  alive   server  0.7.0  2         main
consul-server-3  10.170.131.244:8301  alive   server  0.7.0  2         main
vault-1          10.170.131.69:8301   alive   client  0.7.0  2         main
vault-2          10.170.131.94:8301   alive   client  0.7.0  2         main
```

Repro case:
- Both unsealed
-`-1` is the master.
- step down on `-1`
```
root@vault-1:~# vault status
Sealed: false
Key Shares: 5
Key Threshold: 3
Unseal Progress: 0
Version: 0.6.2
Cluster Name: vault-cluster-7398db7b
Cluster ID: a8cd3f90-1a4c-5e5d-8a64-06af14cead32

High-Availability Enabled: true
	Mode: active
	Leader: http://10.170.131.69:8200

root@vault-1:~# vault step-down

root@vault-1:~# vault status
Sealed: false
Key Shares: 5
Key Threshold: 3
Unseal Progress: 0
Version: 0.6.2
Cluster Name: vault-cluster-7398db7b
Cluster ID: a8cd3f90-1a4c-5e5d-8a64-06af14cead32

High-Availability Enabled: true
	Mode: standby
	Leader: http://10.170.131.94:8200
root@vault-1:~# vault read secret/crash-worker/foo
^C
# Hang
```

On the other node:
```
root@vault-2:~# vault status
Sealed: false
Key Shares: 5
Key Threshold: 3
Unseal Progress: 0
Version: 0.6.2
Cluster Name: vault-cluster-7398db7b
Cluster ID: a8cd3f90-1a4c-5e5d-8a64-06af14cead32

High-Availability Enabled: true
	Mode: standby
	Leader: http://10.170.131.69:8200

# < the other node steps down here >
root@vault-2:~# vault status
Error checking seal status: Get http://127.0.0.1:8200/v1/sys/seal-status: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

The hung vault leader is doing seemingly nothing:
```
root@vault-2:~# ps aux | grep vault
vault     8535  0.0  0.1  37772 22332 ?        Ssl  02:21   0:01 /usr/local/bin/vault server -config=/etc/vault.d
```

strace continues like this:
```
[pid  8536] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  8604] futex(0x1df3998, FUTEX_WAIT, 0, {0, 999900405} <unfinished ...>
[pid  8536] <... select resumed> )      = 0 (Timeout)
[pid  8536] futex(0x1df4058, FUTEX_WAIT, 0, {60, 0} <unfinished ...>
[pid  8538] <... epoll_wait resumed> [{EPOLLIN|EPOLLOUT|EPOLLERR|EPOLLHUP|EPOLLRDHUP, {u32=853363328, u64=140253010414208}}], 128, -1) = 1
[pid  8538] epoll_wait(4,  <unfinished ...>
[pid  8604] <... futex resumed> )       = -1 ETIMEDOUT (Connection timed out)
[pid  8604] futex(0x1df4058, FUTEX_WAKE, 1) = 1
[pid  8536] <... futex resumed> )       = 0
[pid  8604] futex(0xc4203d6910, FUTEX_WAKE, 1) = 1
[pid  8540] <... futex resumed> )       = 0
[pid  8536] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  8604] futex(0x1df4048, FUTEX_WAIT, 0, {0, 100000}) = -1 EAGAIN (Resource temporarily unavailable)
[pid  8540] futex(0x1df4048, FUTEX_WAKE, 1 <unfinished ...>
[pid  8604] epoll_wait(4,  <unfinished ...>
[pid  8540] <... futex resumed> )       = 0
[pid  8604] <... epoll_wait resumed> [], 128, 0) = 0
[pid  8540] futex(0xc4203d6910, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  8536] <... select resumed> )      = 0 (Timeout)
[pid  8604] futex(0xc4203d6910, FUTEX_WAKE, 1 <unfinished ...>
[pid  8540] <... futex resumed> )       = -1 EAGAIN (Resource temporarily unavailable)
[pid  8604] <... futex resumed> )       = 0
[pid  8536] select(0, NULL, NULL, NULL, {0, 20} <unfinished ...>
[pid  8540] futex(0xc4200be510, FUTEX_WAKE, 1) = 1
[pid  8539] <... futex resumed> )       = 0
[pid  8540] futex(0x1df3998, FUTEX_WAIT, 0, {0, 889635971} <unfinished ...>
[pid  8539] futex(0xc4200be510, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  8536] <... select resumed> )      = 0 (Timeout)
[pid  8604] futex(0xc4203d6d10, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid  8536] futex(0x1df4058, FUTEX_WAIT, 0, {60, 0}^Cstrace: Process 8535 detached
```
I tried stopping the old primary `-1` in this case, to see if it held on to the consul lock, but that wasn't the case.

Originally, both were configured to listen on `127.0.0.1`, and I thought it may be an issue related to both masters thinking they were `127.0.0.1`. That doesn't seem to be the case though, as now they are both listening on `0.0.0.0` and see the other's IPv4 addresses.

*Note* configuring `0.0.0.0` actually binds to `::0` or in6_addr_any. [This is a golang thing](https://github.com/golang/go/issues/7411)

I inspected the stack trace to no avail.
I tried getting just `-2` to run, and had trouble.

In syslog, it complained about:
```
Nov 11 03:58:49 vault-2 vault[8958]: 2016/11/11 03:58:49.431173 [ERROR] core: failed to create audit entry: path=file/ error=sanity check failed; unable to open /var/log/vault/audit.log for writing: mkdir /var/log/vault: permission denied
```

created it with:
```
mkdir /var/log/vault
chown vault: /var/log/vault
```
stopped -1
restarted -2
-2 took active in a few seconds
`vault step-down` now works as advertised, and can swap back and forth.

I believe I got into this state because I enabled the file audit on `-1`, and the directory didn't exist on `-2`

Is such a thing possible?

I'm a little surprised that everything *appeared* fine, until it tried to take active.

```
2016/11/11 19:52:27.288735 [WARN ] physical/consul: appending trailing forward slash to path
2016/11/11 19:52:38.270294 [INFO ] core: vault is unsealed
2016/11/11 19:52:38.270349 [WARN ] physical/consul: Concurrent sealed state change notify dropped
2016/11/11 19:52:38.270456 [INFO ] core: entering standby mode


2016/11/11 19:53:46.515299 [INFO ] core: acquired lock, enabling active operation
2016/11/11 19:53:46.633266 [WARN ] physical/consul: Concurrent state change notify dropped
2016/11/11 19:53:46.633296 [INFO ] core: post-unseal setup starting
2016/11/11 19:53:46.639404 [INFO ] core: successfully mounted backend: type=system path=sys/
2016/11/11 19:53:46.639466 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
2016/11/11 19:53:46.639532 [INFO ] core: successfully mounted backend: type=consul path=consul/
2016/11/11 19:53:46.639588 [INFO ] core: successfully mounted backend: type=generic path=secret/
2016/11/11 19:53:46.639651 [INFO ] core: successfully mounted backend: type=mysql path=mysql/
2016/11/11 19:53:46.639712 [INFO ] core: successfully mounted backend: type=mysql path=db/crash-worker/
2016/11/11 19:53:46.639763 [INFO ] core: successfully mounted backend: type=mysql path=db/collab-service/
2016/11/11 19:53:46.639884 [INFO ] rollback: starting rollback manager
2016/11/11 19:53:46.661040 [INFO ] expire: leases restored: restored_lease_count=4
2016/11/11 19:53:46.662942 [ERROR] core: failed to create audit entry: path=file/ error=sanity check failed; unable to open /var/log/vault/audit.log for writing: mkdir /var/log/vault: permission denied
2016/11/11 19:53:46.662973 [INFO ] core: pre-seal teardown starting
2016/11/11 19:53:46.662983 [INFO ] core/stopClusterListener: stopping listeners
```

Appears to be the key.

My interpretation is:
- We're in standby, and unsealed
- the other node steps down, we call `postUnseal`
- `postUnseal` calls `setupAudits`
- `setupAudits` returns an error on *the first* audit backend, before trying others

Note:
```
root@vault-2:~# vault audit-list
Path     Type    Description  Options
file/    file                 file_path=/var/log/vault/audit.log
syslog/  syslog               facility=AUTH tag=vault
```
file is first. *presumably* it is tried in the same order.

- postUnseal then forward the error message:
```golang
/* postUnseal */
1146     if err := c.setupAudits(); err != nil {
1147         return err
1148     }
```

I'm not clear on who is the caller of `postUnseal` though.

It seems to be `runStandby` because we see `core: acquired lock, enabling active operation` in the log, but I don't see that error message:
```
1276         // Attempt the post-unseal process
1277         err = c.postUnseal()
1278         if err == nil {
1279             c.standby = false
1280         }
1281         c.stateLock.Unlock()
1282 
1283         // Handle a failure to unseal
1284         if err != nil {
1285             c.logger.Error("core: post-unseal setup failed", "error", err)
1286             lock.Unlock()
1287             metrics.MeasureSince([]string{"core", "leadership_setup_failed"}, activeTime)
1288             continue
1289         }
```

--- a/vault/cluster.go
+++ b/vault/cluster.go
@@ -316,6 +316,10 @@ func (c *Core) stopClusterListener() {
 		return
 	}
 
+	if !c.clusterListenersRunning {
+		c.logger.Info("core/stopClusterListener: listeners not running")
+		return
+	}
 	c.logger.Info("core/stopClusterListener: stopping listeners")
 
 	// Tell the goroutine managing the listeners to perform the shutdown
@@ -327,6 +331,8 @@ func (c *Core) stopClusterListener() {
 	// bind errors. This ensures proper ordering.
 	c.logger.Trace("core/stopClusterListener: waiting for success notification")
 	<-c.clusterListenerShutdownSuccessCh
+	c.clusterListenersRunning = false
+
 	c.logger.Info("core/stopClusterListener: success")
 }
 

--- a/vault/core.go
+++ b/vault/core.go
@@ -269,6 +269,9 @@ type Core struct {
 	clusterListenerAddrs []*net.TCPAddr
 	// The setup function that gives us the handler to use
 	clusterHandlerSetupFunc func() (http.Handler, http.Handler)
+	// Tracks whether cluster listeners are running, e.g. it's safe to send a
+	// shutdown down the channel
+	clusterListenersRunning bool
 	// Shutdown channel for the cluster listeners
 	clusterListenerShutdownCh chan struct{}
 	// Shutdown success channel. We need this to be done serially to ensure

--- a/vault/request_forwarding.go
+++ b/vault/request_forwarding.go
@@ -145,6 +145,10 @@ func (c *Core) startForwarding() error {
 
 	// This is in its own goroutine so that we don't block the main thread, and
 	// thus we use atomic and channels to coordinate
+	// However, because you can't query the status of a channel, we set a bool
+	// here while we have the state lock to know whether to actually send a
+	// shutdown (e.g. whether the channel will block). See issue #2083.
+	c.clusterListenersRunning = true
 	go func() {
 		// If we get told to shut down...
 		<-c.clusterListenerShutdownCh

